### Exercise 1.1

##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [80]:
import pandas as pd
import numpy as np

trips = pd.read_csv("./2019_Yellow_Taxi_Trip_Data.csv", nrows=5)
trips

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5      1      0.5   
1            11            26             1         10.5      1      0.5   
2           163           162             1          9.5      1      0.5   
3           170           163             1         13.0      1      0.5   
4           163           236             1         10.5      1      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

### Exercise 1.2

##### Find the dimensions (number of rows and number of columns) in the data.

In [81]:
trips.shape

(5, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [82]:
trips.describe(include = "all")

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
count   5.000000                        5                        5   
unique       NaN                        5                        5   
top          NaN  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
freq         NaN                        1                        1   
mean    1.800000                      NaN                      NaN   
std     0.447214                      NaN                      NaN   
min     1.000000                      NaN                      NaN   
25%     2.000000                      NaN                      NaN   
50%     2.000000                      NaN                      NaN   
75%     2.000000                      NaN                      NaN   
max     2.000000                      NaN                      NaN   

        passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
count               5.0       5.000000         5.0                  5   
unique              NaN            NaN         NaN                  1   
top                 NaN            NaN         NaN                  N   
freq                NaN            NaN         NaN                  5   
mean                1.0       2.850000         1.0                NaN   
std                 0.0       2.870697         0.0                NaN   
min                 1.0       1.000000         1.0                NaN   
25%                 1.0       1.360000         1.0                NaN   
50%                 1.0       1.960000         1.0                NaN   
75%                 1.0       2.000000         1.0                NaN   
max                 1.0       7.930000         1.0                NaN   

        pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
count       5.000000      5.000000           5.0     5.000000    5.0      5.0   
unique           NaN           NaN           NaN          NaN    NaN      NaN   
top              NaN           NaN           NaN          NaN    NaN      NaN   
freq             NaN           NaN           NaN          NaN    NaN      NaN   
mean      129.000000    151.400000           1.0    14.600000    1.0      0.5   
std        67.078312     76.608094           0.0     8.429116    0.0      0.0   
min        11.000000     26.000000           1.0     9.500000    1.0      0.5   
25%       138.000000    162.000000           1.0    10.500000    1.0      0.5   
50%       163.000000    163.000000           1.0    10.500000    1.0      0.5   
75%       163.000000    170.000000           1.0    13.000000    1.0      0.5   
max       170.000000    236.000000           1.0    29.500000    1.0      0.5   

        tip_amount  tolls_amount  improvement_surcharge  total_amount  \
count     5.000000      5.000000                    5.0       5.00000   
unique         NaN           NaN                    NaN           NaN   
top            NaN           NaN                    NaN           NaN   
freq           NaN           NaN                    NaN           NaN   
mean      2.960000      1.224000                    0.3      22.58400   
std       3.270505      2.736947                    0.0      14.54879   
min       0.000000      0.000000                    0.3      12.30000   
25%       0.500000      0.000000                    0.3      15.30000   
50%       2.000000      0.000000                    0.3      15.80000   
75%       4.320000      0.000000                    0.3      21.62000   
max       7.980000      6.120000                    0.3      47.90000   

        congestion_surcharge  
count               5.000000  
unique                   NaN  
top                      NaN  
freq                     NaN  
mean                2.000000  
std                 1.118034  
min                 0.000000  
25%                 2.500000  
50%                 2.500000  
75%                 2.500000  
max                 2.500000

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [83]:
fare = (trips.fare_amount.max())
tip = (trips.tip_amount.max())
tolls = (trips.tolls_amount.max())
distance = (trips.trip_distance.max())
print(fare)
print(tip)
print(tolls)
print(distance)

29.5
7.98
6.12
7.93


---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [84]:
meteor = pd.read_csv("./Meteorite_Landings.csv")
meteor = meteor.rename(columns = {'mass (g)':'mass'})
columns_to_drop = ["reclat", "reclong"]
meteor = meteor.drop(columns=columns_to_drop)
meteor = meteor.sort_values(by ='mass', ascending = False)
meteor


name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   
...                ...    ...      ...           ...         ...    ...   
38282   Wei-hui-fu (a)  24231    Valid          Iron         NaN  Found   
38283   Wei-hui-fu (b)  24232    Valid          Iron         NaN  Found   
38285          Weiyuan  24233    Valid  Mesosiderite         NaN  Found   
41472    Yamato 792768  28117    Valid           CM2         NaN  Found   
45698    Zapata County  30393    Valid          Iron         NaN  Found   

                         year             GeoLocation  
16392  01/01/1920 12:00:00 AM   (-19.58333, 17.91667)  
5373   01/01/1818 12:00:00 AM   (76.13333, -64.93333)  
5365   12/22/1575 12:00:00 AM  (-27.46667, -60.58333)  
5370   01/01/1891 12:00:00 AM     (35.05, -111.03333)  
3455   01/01/1898 12:00:00 AM            (47.0, 88.0)  
...                       ...                     ...  
38282  01/01/1931 12:00:00 AM                     NaN  
38283  01/01/1931 12:00:00 AM                     NaN  
38285  01/01/1978 12:00:00 AM   (35.26667, 104.31667)  
41472  01/01/1979 12:00:00 AM       (-71.5, 35.66667)  
45698  01/01/1930 12:00:00 AM           (27.0, -99.0)  

[45716 rows x 8 columns]

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [94]:
just_year = meteor.year.str.slice(6,10)
meteor.replace(to_replace = meteor.year, value= just_year, inplace = True)
meteor

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   
...                ...    ...      ...           ...         ...    ...   
38282   Wei-hui-fu (a)  24231    Valid          Iron         NaN  Found   
38283   Wei-hui-fu (b)  24232    Valid          Iron         NaN  Found   
38285          Weiyuan  24233    Valid  Mesosiderite         NaN  Found   
41472    Yamato 792768  28117    Valid           CM2         NaN  Found   
45698    Zapata County  30393    Valid          Iron         NaN  Found   

                         year             GeoLocation  
16392  01/01/1920 12:00:00 AM   (-19.58333, 17.91667)  
5373   01/01/1818 12:00:00 AM   (76.13333, -64.93333)  
5365   12/22/1575 12:00:00 AM  (-27.46667, -60.58333)  
5370   01/01/1891 12:00:00 AM     (35.05, -111.03333)  
3455   01/01/1898 12:00:00 AM            (47.0, 88.0)  
...                       ...                     ...  
38282  01/01/1931 12:00:00 AM                     NaN  
38283  01/01/1931 12:00:00 AM                     NaN  
38285  01/01/1978 12:00:00 AM   (35.26667, 104.31667)  
41472  01/01/1979 12:00:00 AM       (-71.5, 35.66667)  
45698  01/01/1930 12:00:00 AM           (27.0, -99.0)  

[45716 rows x 8 columns]

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

---

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.